In [156]:
#import library
import pandas as pd
from sqlalchemy import create_engine

In [121]:
#load dataset
data=pd.read_csv('WA_Fn-UseC_-HR-Employee-Attrition.csv',index_col='EmployeeNumber')
# view the first 5 rows
pd.set_option('display.max_columns', None)
print(data.head())

                Age Attrition     BusinessTravel  DailyRate  \
EmployeeNumber                                                
1                41       Yes      Travel_Rarely       1102   
2                49        No  Travel_Frequently        279   
4                37       Yes      Travel_Rarely       1373   
5                33        No  Travel_Frequently       1392   
7                27        No      Travel_Rarely        591   

                            Department  DistanceFromHome  Education  \
EmployeeNumber                                                        
1                                Sales                 1          2   
2               Research & Development                 8          1   
4               Research & Development                 2          2   
5               Research & Development                 3          4   
7               Research & Development                 2          1   

               EducationField  EmployeeCount  EnvironmentSat

In [74]:
#dataset info
data.info()
#Observation
"""HR dataset has 1470 rows and 34 columns.There are no non_null entries in dataset.Each field name is properly written.
Each field has correct dtype. Few fields are ordinal in nature ."""

<class 'pandas.core.frame.DataFrame'>
Index: 1470 entries, 1 to 2068
Data columns (total 34 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   object
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   int64 
 7   EducationField            1470 non-null   object
 8   EmployeeCount             1470 non-null   int64 
 9   EnvironmentSatisfaction   1470 non-null   int64 
 10  Gender                    1470 non-null   object
 11  HourlyRate                1470 non-null   int64 
 12  JobInvolvement            1470 non-null   int64 
 13  JobLevel                  1470 non-null   int64 
 14  JobRole                   147

'HR dataset has 1470 rows and 34 columns.There are no non_null entries in dataset.Each field name is properly written.\nEach field has correct dtype. Few fields are ordinal in nature .'

In [122]:
#check null values 
print(data.isnull().sum())
#Observation
'''Each column has non-null values.'''

Age                         0
Attrition                   0
BusinessTravel              0
DailyRate                   0
Department                  0
DistanceFromHome            0
Education                   0
EducationField              0
EmployeeCount               0
EnvironmentSatisfaction     0
Gender                      0
HourlyRate                  0
JobInvolvement              0
JobLevel                    0
JobRole                     0
JobSatisfaction             0
MaritalStatus               0
MonthlyIncome               0
MonthlyRate                 0
NumCompaniesWorked          0
Over18                      0
OverTime                    0
PercentSalaryHike           0
PerformanceRating           0
RelationshipSatisfaction    0
StandardHours               0
StockOptionLevel            0
TotalWorkingYears           0
TrainingTimesLastYear       0
WorkLifeBalance             0
YearsAtCompany              0
YearsInCurrentRole          0
YearsSinceLastPromotion     0
YearsWithC

'Each column has non-null values.'

In [123]:
#check duplicate
print(data.duplicated().value_counts())
#Observation
'''There is no duplicate value in data.'''

False    1470
Name: count, dtype: int64


'There is no duplicate value in data.'

In [124]:
# check unique values each column
print(data['EmployeeCount'].unique(),
data['BusinessTravel'].unique(),
data['StandardHours'].unique(),
data['JobInvolvement'].unique(),
data['YearsWithCurrManager'].unique())

[1] ['Travel_Rarely' 'Travel_Frequently' 'Non-Travel'] [80] [3 2 4 1] [ 5  7  0  2  6  8  3 11 17  1  4 12  9 10 15 13 16 14]


In [125]:
# remove unvcessary columns
data.drop(columns=['JobInvolvement','YearsWithCurrManager','EmployeeCount','StandardHours','StockOptionLevel'],inplace=True)

In [126]:
#recheck columns
data.columns

Index(['Age', 'Attrition', 'BusinessTravel', 'DailyRate', 'Department',
       'DistanceFromHome', 'Education', 'EducationField',
       'EnvironmentSatisfaction', 'Gender', 'HourlyRate', 'JobLevel',
       'JobRole', 'JobSatisfaction', 'MaritalStatus', 'MonthlyIncome',
       'MonthlyRate', 'NumCompaniesWorked', 'Over18', 'OverTime',
       'PercentSalaryHike', 'PerformanceRating', 'RelationshipSatisfaction',
       'TotalWorkingYears', 'TrainingTimesLastYear', 'WorkLifeBalance',
       'YearsAtCompany', 'YearsInCurrentRole', 'YearsSinceLastPromotion'],
      dtype='object')

In [146]:
#Modify Col Values 'Over18' 
data['Over18']=data['Over18'].map({'Y':'Yes','N': 'No'})

# Convert ALL Ordineal Categories into Labels
'''Attrition
Education
JobSatisfaction
PerformanceRating
RelationshipSatisfaction'''

In [127]:
#Convert Categorical Column into Binary Columns
data['Attrition']=data['Attrition'].map({"Yes":1,"No": 0}).astype(int)

In [128]:
#Convert Numeric Column into Categorical Column
data['Education']=data['Education'].map({1: 'Below College',2:'College',3: 'Bachelor',4: 'Master',5 :'Doctor'}).astype(object)
data['JobSatisfaction']=data['JobSatisfaction'].map({1 :'Low',2 :'Medium',3: 'High',4: 'Very High'}).astype(object)

In [129]:
# drop irrelevent column 'EnvironmentSatisfaction'
data.drop(columns='EnvironmentSatisfaction',inplace=True)

In [133]:
#Convert Numeric Values to Categorical Labels
data['PerformanceRating'] = data['PerformanceRating'].map({
    1: 'Low',
    2: 'Good',
    3: 'Excellent',
    4: 'Outstanding'
}).astype(object)


In [135]:
# convert Numeric Values to Categorical Labels
data['RelationshipSatisfaction'] = data['RelationshipSatisfaction'].map({
    1: 'Low',
    2: 'Medium',
    3: 'High',
    4: 'Very High'
}).astype(object)


In [148]:
# Recheck All Data
data.info()
# get first 5 rows
print(data.head())

<class 'pandas.core.frame.DataFrame'>
Index: 1470 entries, 1 to 2068
Data columns (total 28 columns):
 #   Column                    Non-Null Count  Dtype 
---  ------                    --------------  ----- 
 0   Age                       1470 non-null   int64 
 1   Attrition                 1470 non-null   int64 
 2   BusinessTravel            1470 non-null   object
 3   DailyRate                 1470 non-null   int64 
 4   Department                1470 non-null   object
 5   DistanceFromHome          1470 non-null   int64 
 6   Education                 1470 non-null   object
 7   EducationField            1470 non-null   object
 8   Gender                    1470 non-null   object
 9   HourlyRate                1470 non-null   int64 
 10  JobLevel                  1470 non-null   int64 
 11  JobRole                   1470 non-null   object
 12  JobSatisfaction           1470 non-null   object
 13  MaritalStatus             1470 non-null   object
 14  MonthlyIncome             147

In [150]:
#save Cleaned_hr_data
data.to_csv('cleaned_hr_data.csv',index=True)

In [162]:
#Create Engine and Connect to Database 'hr_attrition_analysis'
engine = create_engine('mysql+pymysql://root:1234@127.0.0.1:3306/hr_attrition_analysis')

In [163]:
#Connect to SQL Database
#View table names
df=pd.read_sql('Show Tables',engine)
df

Tables_in_hr_attrition_analysis
0                 cleaned_hr_data

In [164]:
#View first 5 records
df=pd.read_sql('select * from cleaned_hr_data limit 5',engine)
df

EmployeeNumber  Age  Attrition     BusinessTravel  DailyRate  \
0               1   41          1      Travel_Rarely       1102   
1               2   49          0  Travel_Frequently        279   
2               4   37          1      Travel_Rarely       1373   
3               5   33          0  Travel_Frequently       1392   
4               7   27          0      Travel_Rarely        591   

               Department  DistanceFromHome      Education EducationField  \
0                   Sales                 1        College  Life Sciences   
1  Research & Development                 8  Below College  Life Sciences   
2  Research & Development                 2        College          Other   
3  Research & Development                 3         Master  Life Sciences   
4  Research & Development                 2  Below College        Medical   

   Gender  HourlyRate  JobLevel                JobRole JobSatisfaction  \
0  Female          94         2        Sales Executive       Very High   
1    Male          61         2     Research Scientist          Medium   
2    Male          92         1  Laboratory Technician            High   
3  Female          56         1     Research Scientist            High   
4    Male          40         1  Laboratory Technician          Medium   

  MaritalStatus  MonthlyIncome  MonthlyRate  NumCompaniesWorked Over18  \
0        Single           5993        19479                   8    Yes   
1       Married           5130        24907                   1    Yes   
2        Single           2090         2396                   6    Yes   
3       Married           2909        23159                   1    Yes   
4       Married           3468        16632                   9    Yes   

  OverTime  PercentSalaryHike PerformanceRating RelationshipSatisfaction  \
0      Yes                 11         Excellent                      Low   
1       No                 23       Outstanding                Very High   
2      Yes                 15         Excellent                   Medium   
3      Yes                 11         Excellent                     High   
4       No                 12         Excellent                Very High   

   TotalWorkingYears  TrainingTimesLastYear  WorkLifeBalance  YearsAtCompany  \
0                  8                      0                1               6   
1                 10                      3                3              10   
2                  7                      3                3               0   
3                  8                      3                3               8   
4                  6                      3                3               2   

   YearsInCurrentRole  YearsSinceLastPromotion  
0                   4                        0  
1                   7                        1  
2                   0                        0  
3                   7                        3  
4                   2                        2

In [174]:
#Show Attrition Rate of Company('Yes' =1 ,'No' = 0)
df=pd.read_sql('select Attrition, count(Attrition) as Attrition_Rate from cleaned_hr_data group by Attrition' ,engine)
df

Attrition  Attrition_Rate
0          1             237
1          0            1233

#Observation
Attrition Status	Count
Yes (1)	            237
No (0)	           1233
"Out of 1,470 employees, 237 have left the company while 1,233 have remained."

In [183]:
#Show Attrition Rate by Age Grouping
df=pd.read_sql('''with cte as (select Case when Age < 18 then "Minor" when  Age between 18 and 30 then  "Young Adult" when Age
between 31 and 45 then "Middle-Aged" when Age >=46 then "Older" end  as Age_Grouping,Attrition from cleaned_hr_data)   select age_grouping,
count(Attrition) as Attrition_Rate from cte  where Attrition = 1  group by age_grouping''' ,engine)
df

age_grouping  Attrition_Rate
0  Middle-Aged             103
1  Young Adult             100
2        Older              34

#Observation

age_grouping	Attrition_Rate
0	Middle-Aged	  103
1	Young Adult	  100
2	Older         34
"""Middle-aged employees have the highest number of attritions. Older employees appear to be more loyal to the company compared 
to younger and middle-aged groups."""



In [188]:
#Show Attrition Rate by Department
df=pd.read_sql('''select Department, count(Attrition) as Attrition_Rate from cleaned_hr_data where Attrition = 1 group by Department
order by Attrition_Rate desc''' ,engine)
df

Department  Attrition_Rate
0  Research & Development             133
1                   Sales              92
2         Human Resources              12

#Observation
     Department	                  Attrition_Rate
0	Research & Development	        133
1	Sales	                         92
2	Human Resources	                 12
"The Research & Development department has a significantly higher attrition count compared to Sales and Human Resources. The company should consider conducting a survey to assess factors such as job satisfaction, salary structure, and work environment to identify the root causes of attrition."

In [189]:
#Attrition by Gender
df=pd.read_sql('''select Gender, count(Attrition) as Attrition_Rate from cleaned_hr_data where Attrition = 1 group by Gender
order by Attrition_Rate desc''' ,engine)
df

Gender  Attrition_Rate
0    Male             150
1  Female              87

#Observation
Gender	Attrition_Rate
0	Male	150
1	Female	87
"Male employees show a higher attrition rate than female employees."

In [199]:
# Income Group vs Attrition Rate
df=pd.read_sql('''with cte as (select case when MonthlyIncome < 5000 then "Low" when MonthlyIncome between 5000 and 15000 then 'Medium'
when MonthlyIncome > 15001 then
'High' end as Income_Group,Attrition  from cleaned_hr_data) select Income_Group ,count(Attrition) as Attrition_Rate from cte where Attrition = 1 group by
Income_Group order by Attrition_Rate desc''',engine)
df

Income_Group  Attrition_Rate
0          Low             163
1       Medium              69
2         High               5

#Observation
	Income_Group	Attrition_Rate
0	Low	            163
1	Medium	        69
2	High	        5
"Employees in the Low Income Group (< $5,000) have the highest attrition rate. Lower income is often linked to higher unemployment rates."

In [210]:
#Count of employees vs monthly income (grouped)
df=pd.read_sql('''with cte as (select case when MonthlyIncome < 5000 then "Low" when MonthlyIncome between 5000 and 15000 then 'Medium'
when MonthlyIncome > 15001 then
'High' end as Income_Group,EmployeeNumber  from cleaned_hr_data) select Income_Group ,count(EmployeeNumber) as Total_Employee from cte
group by Income_Group Order by Total_Employee desc''',
               engine)
df

Income_Group  Total_Employee
0          Low             749
1       Medium             588
2         High             133

#Observation
Income_Group	Total_Employee
0	Low	         749
1	Medium	     588
2	High	     133
"The majority of employees (749) fall into the Low Income Group, followed by 588 in the Medium group, and only 133 in the High Income Group."

In [217]:
#Income_Group vs 
df=pd.read_sql('''with cte as (select case when MonthlyIncome < 5000 then "Low" when MonthlyIncome between 5000 and 15000 then 'Medium'
when MonthlyIncome > 15001 then
'High' end as Income_Group,EmployeeNumber,Attrition  from cleaned_hr_data) 
select Income_Group ,count(EmployeeNumber) as Attrited_Employees,
(COUNT(*) * 100.0 / 1470) as Attrition_Rate from cte where Attrition = 1
group by Income_Group Order by Attrited_Employees desc''',
               engine)
df

Income_Group  Attrited_Employees  Attrition_Rate
0          Low                 163        11.08844
1       Medium                  69         4.69388
2         High                   5         0.34014

#Observation
	Income_Group Attrited_Employees	Attrition_Rate
0	Low	             163	         11.08844
1	Medium	         69	              4.69388
2	High	         5	              0.34014


'Employees in the Low-Income group have the highest attrition rate at 11.08%, compared to 4.69% in the Medium-Income group.'

In [223]:
#Job Role vs. Average Monthly Income 
df=pd.read_sql('select JobRole,round(avg( MonthlyIncome ),2) as AVG_Monthly_Income from cleaned_hr_data group by JobRole order by AVG_Monthly_Income desc ',engine)
df

JobRole  AVG_Monthly_Income
0                    Manager            17181.68
1          Research Director            16033.55
2  Healthcare Representative             7528.76
3     Manufacturing Director             7295.14
4            Sales Executive             6924.28
5            Human Resources             4235.75
6         Research Scientist             3239.97
7      Laboratory Technician             3237.17
8       Sales Representative             2626.00

#Observation
     JobRole	                AVG_Monthly_Income
0	Manager	                    17181.68
1	Research Director	        16033.55
2	Healthcare Representative	7528.76
3	Manufacturing Director	    7295.14
4	Sales Executive         	6924.28
5	Human Resources         	4235.75
6	Research Scientist      	3239.97
7	Laboratory Technician	    3237.17
8	Sales Representative	    2626.00
'Managers have the highest average monthly income at $17,181, while Sales Representatives earn the lowest at $2,626.'

In [238]:
# Job Level vs. Job Satisfaction 
df=pd.read_sql('select 	JobRole,JobSatisfaction from cleaned_hr_data ',engine)
pd.crosstab(index=df['JobRole'],columns=df['JobSatisfaction'],margins=True)

JobSatisfaction            High  Low  Medium  Very High   All
JobRole                                                      
Healthcare Representative    43   26      19         43   131
Human Resources              13   10      16         13    52
Laboratory Technician        75   56      48         80   259
Manager                      27   21      21         33   102
Manufacturing Director       49   26      32         38   145
Research Director            27   15      16         22    80
Research Scientist           90   54      53         95   292
Sales Executive              91   69      54        112   326
Sales Representative         27   12      21         23    83
All                         442  289     280        459  1470

#Observation
JobSatisfaction	            High	Low	 Medium	Very High	All
JobRole					
Healthcare Representative	43	   26	19	43	131
Human Resources	            13	  10	16	13	52
Laboratory Technician	    75	  56	48	80	259
Manager                 	27	  21	21	33	102
Manufacturing Director	   49	  26	32	38	145
Research Director	       27	  15	16	22	80
Research Scientist	       90	  54	53	95	292
Sales Executive            91	  69	54	112	326
Sales Representative	   27	  12	21	23	83
 All	                  442	  289	280	459	1470
'''Among the total 1,470 employees:

459 employees are very highly satisfied

442 employees are highly satisfied

Sales Executives, Research Scientists, and Laboratory Technicians show a wide spread of employees reporting both low and very high job satisfaction possibly due to varying personal interactions with the working environment.''' 

In [245]:
# Job Role vs. Average Years in Company
df=pd.read_sql('select JobRole,round(avg(YearsAtCompany),0) as Avg_Year from cleaned_hr_data group by JobRole order by Avg_Year ',engine)
df


JobRole  Avg_Year
0       Sales Representative       3.0
1         Research Scientist       5.0
2      Laboratory Technician       5.0
3            Human Resources       5.0
4            Sales Executive       7.0
5     Manufacturing Director       8.0
6  Healthcare Representative       8.0
7          Research Director      11.0
8                    Manager      14.0

#Observation
JobRole	Avg_Year
0	Sales Representative	3.0
1	Research Scientist	5.0
2	Laboratory Technician	5.0
3	Human Resources	5.0
4	Sales Executive	7.0
5	Manufacturing Director	8.0
6	Healthcare Representative	8.0
7	Research Director	11.0
8	Manager	14.0
'''Managers and Research Directors show the highest retention. Sales Representatives stay the shortest, 
indicating possible dissatisfaction or high turnover. The company should investigate the reasons behind this.'''

In [253]:
# Training vs Average Salary Hike
df=pd.read_sql('select TrainingTimesLastYear, round(avg(PercentSalaryHike),2) as Salary_Incr_Per from cleaned_hr_data group by  TrainingTimesLastYear ',engine)
df

TrainingTimesLastYear  Salary_Incr_Per
0                      0            15.37
1                      3            15.30
2                      2            15.18
3                      5            15.21
4                      1            15.15
5                      4            14.93
6                      6            15.22

#Observation
	TrainingTimesLastYear	Salary_Incr_Per
0	      0	                15.37
1	      3	                15.30
2	      2             	15.18
3	      5	                15.21
4	      1              	15.15
5	      4	                14.93
6	      6	                15.22

"Training frequency does not show a significant impact on average salary hike. Both trained and untrained employees received similar salary increases."

In [ ]:
Sumamary:
Attrition rate: 16.12%, higher among young and middle-aged employees.

More males (150) left than females (87).

R&D department shows the highest attrition.

11% of low-income employees (≤ $5,000) left the company.

Most employees fall in the low-income group (749).

Managers and Research Directors earn the most; Sales Reps the least.

Job satisfaction is mostly high or very high (442 + 459 employees).

Managers and Research Directors stay the longest; Sales Reps, the shortest.

Training has no major impact on salary hikes.


